crates/ratchet-models/src/moondream/model.rs

```diff
@@ -0,0 +1,282 @@
+                                Some(lt(&format!("vision_encoder.encoder.model.visual.blocks.{}.mlp.fc1.bias", layer))),
+                            ),
+                            Linear::new(
+                                lt(&format!("vision_encoder.encoder.model.visual.blocks.{}.mlp.fc2.weight", layer)),
+                                Some(lt(&format!("vision_encoder.encoder.model.visual.blocks.{}.mlp.fc2.bias", layer))),
+                            ),
+                        ),
+                        LayerNorm::new(
+                            lt(&format!("vision_encoder.encoder.model.visual.blocks.{}.norm1.weight", layer)),
+                            Some(lt(&format!("vision_encoder.encoder.model.visual.blocks.{}.norm1.bias", layer))),
+                            ln_eps,
+                        ),
+                        LayerNorm::new(
+                            lt(&format!("vision_encoder.encoder.model.visual.blocks.{}.norm2.weight", layer)),
+                            Some(lt(&format!("vision_encoder.encoder.model.visual.blocks.{}.norm2.bias", layer))),
+                            ln_eps,
+                        ),
+                    )
+                }).collect::<Vec<_>>(),
+            LayerNorm::new(lt("vision_encoder.encoder.model.visual.norm.weight"), Some(lt("vision_encoder.encoder.model.visual.norm.bias")), ln_eps),
+        );
+
+        let vision_encoder = VisionEncoder::new(projection, transformer);
+        Ok(Self {
+            vision_encoder,
+            text_model,
+        })
+    }
+}
+
+#[cfg(all(test, not(target_arch = "wasm32")))]
+mod tests {
+    use std::fs;
+
+    use anyhow::Ok;
+    use hf_hub::api::sync::Api;
+    use ratchet::{shape, test_util::run_py_prg, Device, DeviceRequest, Tensor};
+    use ratchet_loader::gguf;
+    use ratchet_nn::Module;
+    use tokenizers::Tokenizer;
+
+    use crate::moondream::{
+        generate::generate, text_model::TextModel, vision_encoder::VisionEncoder,
+    };
+
+    use super::Moondream;
+
+    fn vision_ground_truth(tensor: Tensor) -> anyhow::Result<Tensor> {
+        let prg = r#"
+from transformers import AutoModelForCausalLM, AutoTokenizer
+import torch
+
+def ground(*args):
+    tensor = torch.from_numpy(args[0])
+    model_id = "vikhyatk/moondream2"
+    revision = "2024-05-20"
+    model = AutoModelForCausalLM.from_pretrained(
+        model_id, trust_remote_code=True, revision=revision
+    )
+    return model.encode_image(tensor).numpy()
+"#;
+
+        run_py_prg(prg.to_string(), &[&tensor], &[], ratchet::DType::F32)
+    }
+
+    #[test]
+    fn vision_encoder() {
+        let device = Device::request_device(DeviceRequest::GPU).unwrap();
+
+        let api = Api::new().unwrap();
+        let model_repo = api.model("tgestson/ratchet-moondream2".to_string());
+        let model_path = model_repo.get("moondream_f32.gguf").unwrap();
```

Comment:
Changing this to be smaller might also help fix the test, the metal runners on CI are weak.

Reply:
Unfortunately we won't get a close match with the PyTorch ground truth that operates on f16 if we use the quantized weights.  These gguf weights are already stored in f16 despite the naming.  Maybe for now we use
`#[cfg_attr(feature = "ci", ignore)]`?

Reply:
@timgestson Agreed